File: KinetisList-iOS/KinetisList/KinetisList/ParseXmlResource.swift
//
//  ParseXmlResource.swift
//  KinetisList
//
//  Created by MAD-Test on 15-1-22.
//  Copyright (c) 2015年 FSL. All rights reserved.
//

import Foundation
public class ParseXmlResource{
    let FEATURES:String = "FEATURES"
    let HEADER:String = "HEADER"
    let FAMILY:String = "FAMILY"
    let DEVICE:String = "DEVICE"
    let ITEM:String = "ITEM"
    
    let HEADER_NAME:String = DBdefine.Header().COLUMN_NAME_NAME
    let HEADER_TYPE:String = DBdefine.Header().COLUMN_NAME_TYPE
    let HEADER_WIDTH = DBdefine.Header().COLUMN_NAME_WIDTH
    let HEADER_SHOW = DBdefine.Header().COLUMN_NAME_SHOW
    
    let ATTR_Name:String = "Name"
    let ATTR_Data:String = "Data"
    
    var itemIndex:Int = 0
    var NumColumns:Int = 0
    var HeaderMap = Dictionary<String,Int>()
    
    var xDeviceRow:[String] = [String]()
    var familyName:String?

    var xmls:[XMLIndexer] = [XMLIndexer]()
    var sourceName:[String] = ["Hot parts", "Conditional parts", "Legacy parts"]
    init(xmlfiles:String...){
        for xmlfile in xmlfiles{
            self.xmls.append(SWXMLHash.parse(xmlfile))
        }
    }
    
    func getData(inout xmlHeader:Array<DBdefine.HeaderItem>, inout xmlDevice:[[String]]){
        //parse header
        var i:Int = 0;
        for xml in xmls{
            for head in xml[FEATURES][HEADER][ITEM]{
                //var Header_Name = head[ITEM].element!.attributes["Width"]!
                var xHeader = DBdefine.HeaderItem()
                xHeader.Name  = head.element!.attributes[HEADER_NAME]!
                xHeader.Title = head.element!.attributes[ATTR_Data]!
                xHeader.Type = head.element!.attributes[HEADER_TYPE]!
                xHeader.width = head.element!.attributes[HEADER_WIDTH]!.toInt()!
                if((head.element!.attributes[HEADER_SHOW]) != nil && head.element!.attributes[HEADER_SHOW] == "false"){
                    xHeader.Visible = false
                }
                HeaderMap[xHeader.Name] = itemIndex
                xmlHeader.append(xHeader)
                itemIndex++
            }
            //parse devices
            xDeviceRow = [String](count: xmlHeader.count, repeatedValue: (""))
            for family in xml[FEATURES][FAMILY]{
                familyName = family.element!.attributes[ATTR_Name]
                for devices in family[DEVICE]{
                    var column:Int? = HeaderMap["SubFamily"]
                    if((column) != nil){
                        xDeviceRow[column!] = familyName!
                    }
                    for items in devices[ITEM]{
                        var itemName:String = items.element!.attributes[ATTR_Name]!
                        column = HeaderMap[itemName]
                        if((column) != nil){
                            xDeviceRow[column!] = items.element!.attributes[ATTR_Data]!
                        }
                    }
                    xDeviceRow[xDeviceRow.count] = sourceName[i]
                    xmlDevice.append(xDeviceRow)
                }
            }
        }
        i++
    }
    
    
}

File: KinetisList/app/src/main/java/com/fsl/mcu/kinetislist/ParseXmlResource.java
package com.fsl.mcu.kinetislist;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import android.content.res.XmlResourceParser;
import android.util.Log;

public class ParseXmlResource {
	private static final String TAG = "ParseXmlResource";

	private static final String	FEATURES = "FEATURES";
	private static final String	HEADER = "HEADER";
	private static final String	FAMILY = "FAMILY";
	private static final String	DEVICE = "DEVICE";
	private static final String	ITEM = "ITEM";

	private static final String	HEADER_NAME = DBdefine.Header.COLUMN_NAME_NAME;
	private static final String	HEADER_TITLE = DBdefine.Header.COLUMN_NAME_TITLE;
	private static final String	HEADER_TYPE = DBdefine.Header.COLUMN_NAME_TYPE;
	private static final String	HEADER_WIDTH = DBdefine.Header.COLUMN_NAME_WIDTH;
	private static final String	HEADER_SHOW = DBdefine.Header.COLUMN_NAME_SHOW;
	
	private static final String	ATTR_Name = "Name";
	private static final String	ATTR_Data = "Data";

	private boolean decode_header = false;
	private boolean decode_device = false;
	private int itemIndex=0;
	
	private int NumColumns = 0;
	private final HashMap<String, Integer> HeaderMap = new HashMap<String, Integer>();
	
	private String[] xDeviceRow = null;
	private String familyName;
	
	private static XmlResourceParser[] xmlParsers;
	public ParseXmlResource(XmlResourceParser... xmlParsers) {
		ParseXmlResource.xmlParsers = xmlParsers;
		
	}

	public void getData(ArrayList<DBdefine.xHeaderItem> xmlHeader, ArrayList<String[]> xmlDevice)
				throws XmlPullParserException, IOException
	{
	for (XmlResourceParser xmlParser: xmlParsers){
		familyName = new String();
        String SourceName = new String();
		int eventType = xmlParser.getEventType();
		String thisTag;
		while (eventType != XmlPullParser.END_DOCUMENT) {
			switch (eventType) {
			case XmlPullParser.START_DOCUMENT:
				break;
			case XmlPullParser.START_TAG:
				thisTag = xmlParser.getName();
				if (thisTag.equals(ITEM)) {
					if (decode_device) {
						String itemName = xmlParser.getAttributeValue(null, ATTR_Name);
						Integer column = HeaderMap.get(itemName);
						if (column != null)
							xDeviceRow[column.intValue()] = xmlParser.getAttributeValue(null, ATTR_Data);
					}
					else if (decode_header) {
						DBdefine.xHeaderItem xHeader = new DBdefine.xHeaderItem();

						xHeader.Name = xmlParser.getAttributeValue(null, HEADER_NAME);
						xHeader.Title = xmlParser.getAttributeValue(null, ATTR_Data);	// COLUMN_NAME_TITLE
						xHeader.Type = xmlParser.getAttributeValue(null, HEADER_TYPE);
						String str = xmlParser.getAttributeValue(null, HEADER_WIDTH);
						try {
							xHeader.Width = Integer.valueOf(str.trim());
						} catch (NumberFormatException eNumber) {
							xHeader.Width = Integer.valueOf(160);	// default value
						}
                        String visible = xmlParser.getAttributeValue(null, HEADER_SHOW);
                        if (visible == null) {
                            xHeader.Visible = true;
                        }else {
                            xHeader.Visible = Boolean.valueOf(visible);
                        }
						if(!HeaderMap.containsKey(xHeader.Name)){
							HeaderMap.put(xHeader.Name, itemIndex);	// for creating the device row
							xmlHeader.add(xHeader);
						}
					}
					itemIndex++;
					break;
				}
				if (thisTag.equals(DEVICE)) {
					decode_device = true;
					itemIndex = 0;
			// Ignore the device name	
			//		xmlParser.getAttributeValue(null, ATTR_Name);
					xDeviceRow = new String[NumColumns+1];
					xDeviceRow[NumColumns] = SourceName;
					Integer column = HeaderMap.get("SubFamily");
					if (column != null)
						xDeviceRow[column] = familyName;
					break;
				}
				if (thisTag.equals(FAMILY)) {
					familyName = xmlParser.getAttributeValue(null, ATTR_Name);
					break;
				}
				if (thisTag.equals(HEADER)) {
					decode_header = true;
					itemIndex = 0;
					break;
				}
				if (thisTag.equals(FEATURES)) {
                    SourceName = xmlParser.getAttributeValue(null, ATTR_Name);
                    break;
                }
				break;
			case XmlPullParser.END_TAG:
				thisTag = xmlParser.getName();
				if (thisTag.equals(HEADER)) {
					decode_header = false;
					NumColumns = itemIndex;
					break;
				}
				if (thisTag.equals(DEVICE)) {
					decode_device = false;
					xmlDevice.add(xDeviceRow);
					break;
				}
				if (thisTag.equals(FAMILY)) {
					familyName = null;
				}	
				break;
			case XmlPullParser.TEXT:
				System.out.println("Text "+ xmlParser.getText());
				break;
			}
			
			eventType = xmlParser.next();
		}
    }
	}
}
